packages/features/pbac/utils/permissionTraversal.ts

```diff
@@ -0,0 +1,103 @@
+import { CrudAction, type CustomAction, PERMISSION_REGISTRY } from "../domain/types/permission-registry";
+
+/**
+ * Generic permission graph traversal function using BFS
+ * @param startPermission The permission to start traversal from
+ * @param direction Whether to find dependencies or dependents
+ * @returns Array of related permissions (excluding the start permission itself)
+ */
+export const traversePermissions = (
+  startPermission: string,
+  direction: "dependencies" | "dependents"
+): string[] => {
+  const visited = new Set<string>();
+  const result = new Set<string>();
+  const queue: string[] = [startPermission];
+
+  while (queue.length > 0) {
+    const currentPermission = queue.shift()!;
```

Comment:
This should still be fine, but just to be safer...

```suggestion
    const currentPermission = queue.shift();
    if (!currentPermission) {
      break;
    }
```